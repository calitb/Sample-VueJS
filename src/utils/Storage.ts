export function removeItem(key: string) {
  sessionStorage.removeItem(key);
}

export function getItem(key: string): string | null {
  return sessionStorage.getItem(key);
}

export function setItem(key: string, value: string) {
  sessionStorage.setItem(key, value);
}
