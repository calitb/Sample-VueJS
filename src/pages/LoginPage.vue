<template>
  <div class="container">
    <b-form>
      <b-row>
        <b-col>Username</b-col>
        <b-col> <b-form-input v-model="username" /> </b-col>
      </b-row>

      <b-row>
        <b-col>Password</b-col>
        <b-col> <b-form-input type="password" v-model="password" /> </b-col>
      </b-row>

      <b-row>
        <b-col>
          <b-button block variant="primary" :disabled="!validated" @click="login">Login</b-button>
        </b-col>
      </b-row>
    </b-form>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { setItem } from '@/utils/Storage';

export default Vue.extend({
  name: 'LoginPage',
  data() {
    return {
      username: '',
      password: ''
    };
  },
  computed: {
    validated(): boolean {
      return this.username.length > 0 && this.password.length > 0;
    }
  },
  methods: {
    login(): void {
      setItem('SESSION', 'true');
      this.$router.push({ name: 'list' });
    }
  }
});
</script>

<style scoped>
.container {
  margin: 10px;
  padding: 5px;
  border: 1px solid black;
}
</style>
