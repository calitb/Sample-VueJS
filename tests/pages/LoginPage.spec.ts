import { shallowMount, mount, Wrapper } from '@vue/test-utils';
import LoginPage from '@/pages/LoginPage.vue';

import * as Storage from '@/utils/Storage';

import { createLocalVue } from '../utils';
const localVue = createLocalVue();

describe('Login Page', () => {
  describe('Empty form', () => {
    const wrapper = shallowMount(LoginPage, { localVue });

    it('should render the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render a form', () => {
      const form = wrapper.find('b-form-stub');
      expect(form.exists()).toBeTruthy();
    });

    it('should render two inputs', () => {
      const form = wrapper.find('b-form-stub');
      const inputs = form.findAll('b-form-input-stub');
      expect(inputs).toHaveLength(2);
    });

    it('should render a disabled button', () => {
      const form = wrapper.find('b-form-stub');
      const button = form.find('b-button-stub');
      expect(button.exists()).toBeTruthy();

      expect(button.attributes().disabled).toBe('true');
    });

    it('should not render a message', () => {
      const message = wrapper.find('p');
      expect(message.exists()).toBeFalsy();
    });
  });

  describe('Filled form', () => {
    const wrapper = shallowMount(LoginPage, { localVue });

    beforeAll(async () => {
      await wrapper.setData({
        username: 'asas',
        password: '12345'
      });
    });

    it('should render the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render filled fields', () => {
      const form = wrapper.find('b-form-stub');
      const inputs = form.findAll('b-form-input-stub');

      expect(inputs.at(0).attributes().value).toBe('asas');
      expect(inputs.at(1).attributes().value).toBe('12345');
    });

    it('should render an enable button', () => {
      const form = wrapper.find('b-form-stub');
      const button = form.find('b-button-stub');
      expect(button.exists()).toBeTruthy();

      expect(button.attributes().disabled).toBe(undefined);
    });

    it('should not render a message', () => {
      const message = wrapper.find('p');
      expect(message.exists()).toBeFalsy();
    });
  });

  describe('Events', () => {
    let spySetItem: jest.SpyInstance<void, [string, string]>;
    let wrapper: Wrapper<any>;

    const localVue = createLocalVue();

    const pushHandler = jest.fn();
    const $router = {
      push: pushHandler
    };

    beforeAll(async () => {
      spySetItem = jest.spyOn(Storage, 'setItem').mockImplementation((key: string) => {});
      wrapper = mount(LoginPage, {
        localVue,
        mocks: {
          $router
        }
      });

      await wrapper.setData({
        username: 'asas',
        password: '12345'
      });
    });

    afterAll(() => {
      spySetItem.mockRestore();
    });

    it('should perform the login when button is clicked', () => {
      const button = wrapper.find('button');
      expect(button.exists()).toBeTruthy();

      button.trigger('click');

      expect(pushHandler).toBeCalledTimes(1);
      expect(pushHandler).toBeCalledWith({ name: 'list' });

      expect(spySetItem).toBeCalledTimes(1);
      expect(spySetItem).toBeCalledWith('SESSION', 'true');
    });
  });
});
