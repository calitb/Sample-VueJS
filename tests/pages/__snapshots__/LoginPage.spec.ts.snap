// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Empty form should render the component 1`] = `
<div
  class="container"
>
  <b-form-stub>
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        Username
      </b-col-stub>
       
      <b-col-stub
        tag="div"
      >
        <b-form-input-stub
          debounce="0"
          type="text"
          value=""
        />
      </b-col-stub>
    </b-row-stub>
     
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        Password
      </b-col-stub>
       
      <b-col-stub
        tag="div"
      >
        <b-form-input-stub
          debounce="0"
          type="password"
          value=""
        />
      </b-col-stub>
    </b-row-stub>
     
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        <b-button-stub
          block="true"
          disabled="true"
          tag="button"
          target="_self"
          type="button"
          variant="primary"
        >
          Login
        </b-button-stub>
      </b-col-stub>
    </b-row-stub>
  </b-form-stub>
</div>
`;

exports[`Login Page Filled form should render the component 1`] = `
<div
  class="container"
>
  <b-form-stub>
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        Username
      </b-col-stub>
       
      <b-col-stub
        tag="div"
      >
        <b-form-input-stub
          debounce="0"
          type="text"
          value="asas"
        />
      </b-col-stub>
    </b-row-stub>
     
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        Password
      </b-col-stub>
       
      <b-col-stub
        tag="div"
      >
        <b-form-input-stub
          debounce="0"
          type="password"
          value="12345"
        />
      </b-col-stub>
    </b-row-stub>
     
    <b-row-stub
      tag="div"
    >
      <b-col-stub
        tag="div"
      >
        <b-button-stub
          block="true"
          tag="button"
          target="_self"
          type="button"
          variant="primary"
        >
          Login
        </b-button-stub>
      </b-col-stub>
    </b-row-stub>
  </b-form-stub>
</div>
`;
